[inductor] Enable floor_div indexing to work under ABI-compat mode

int3 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113276

Previously, floor_div operations were defined in
ATen/native/BinaryOps.h. Since this header was not included under
ABI-compat mode, trying to use those indexing operations would result in
compilation errors.

Technically, it is safe to use aten::native::floor_div_* functions in
ABI-compat mode as they are header-only; we could simply include
BinaryOps.h. However, there are other declarations in BinaryOps.h that
are not binary-compatible, so this is not ideal. Thus, I have moved those
functions into a separate file, and put them under c10/util, since they
don't really have tensor-specific logic.

c10 functions are not all header-only, so this still isn't ideal, but
this still seems like an improvement. Moreover, cpp_prefix.h -- used
when compiling cpp kernels -- already includes c10 header files, so
ABI-compatibility already depends on maintaining some c10 functions as
header-only.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

int3 committed:
[inductor] Add abi-compatible shims for div_floor_* functions

We need these in order to compile indexing expressions.

[ghstack-poisoned]

int3 commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -55,6 +56,14 @@ c10::optional<T> pointer_to_optional(U* ptr) {
 
 } // namespace
 
+int64_t aoti_torch_div_floor_int64(int64_t a, int64_t b) {
```
it appears we only use `div_floor_integer<int64_t>` when emitting indexing expressions, so I only created this particular instantiation (since C doesn't support overloading)

desertfire replied:
I think I was doing the same thing for `div_floor_integer` at some point, but realized we need a way to capture exception here, like `divide-by-zero`. So making the API similar to others here by using `AOTI_TORCH_CONVERT_EXCEPTION_TO_ERROR_CODE` seems inevitable, unless @ezyang has a better idea.

ezyang replied:
@swolchok if you're doing a CPU model and there's some code where we do an integer divide, which potentially may have a zero denominator. What would you prefer the AOTInductor DSO behavior to be in this case?

swolchok replied:
I don't understand -- div_floor_integer doesn't check for divide by zero.

I also want to clear up what is supposed to happen -- integer division by zero has undefined behavior (https://en.cppreference.com/w/c/language/operator_arithmetic#Division). it is not guaranteed to crash the process. it is very unlikely to raise a C++ exception. C++ authors must guarantee that their code does not perform integer division by zero.

swolchok replied:
supposing that div_floor_integer *did* check for divide by zero, I would prefer that it was compiled into the DSO, not going through an aoti_torch shim, because those are relatively expensive and it is a small function.

swolchok replied:
in particular we could simply emit a zero check for the denominator inline and throw an exception, which would be translated into an error code at the boundary and then back into an exception on the other side

int3 replied:
I guess regardless of whether it does check or not, it's better to have it compiled into the DSO and/or inlined? Let me look into that...

swolchok replied:
> compiled into the DSO and/or inlined?

can't inline across a DSO boundary, gotta compile into the DSO

int3 replied:
Yeah I didn't really phrase things correctly... should not have said "or". I just meant that if we compile the `div_floor_*` functions into the DSO (I assume that means model.so), then they could possibly be inlined into their callers.

int3 commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -11,6 +11,7 @@
 #include <cstdio>
 #include <iostream>
 #include <memory>
+#include <ATen/native/BinaryOps.h>
```
not sure if this include is OK, or if I should make it part of `AT_PER_OPERATOR_HEADERS`?

int3 committed:
Update on "[inductor] Add abi-compatible shims for div_floor_* functions"


We need these in order to compile indexing expressions.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

chenyang78 reviewed (approved):
LGTM. Thanks!

int3 committed:
Update on "[inductor] Add abi-compatible shims for div_floor_* functions"


We need these in order to compile indexing expressions.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Add abi-compatible shims for div_floor_* functions"


We need these in order to compile indexing expressions.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Add torch::inductor::div_floor_{int64, double}"


These functions are needed for computing indexing expressions.

We were previously calling into
`aten::native::div_floor_{integer,floating}` when compiling the cpp
wrapper code in non-ABI-compatible mode. However, these functions were
not available when compiling under ABI-compatible mode as we did not
include ATen/native/BinaryOps.h. It is technically safe to call those
functions in ABI-compatible mode as they are header-only, but I didn't
want to include BinaryOps.h into the cpp wrapper file because it exposes
other potentially non-ABI-stable declarations.

As such, I have created a new torch/csrc/inductor/math.h file that
contains versions of div_floor_* that are clearly uncoupled from libc10
and libtorch. We only need to care about int64_t and double types when
computing indexing expressions, so I have only defined
div_floor_{int64_, double}. This makes it easy to decouple from c10
code.

I also noticed that when compiling the cpp kernels, we include
BinaryOps.h directly. I have swapped that out for the new math.h to
better enforce the ABI-compatible interface.

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Add torch::inductor::div_floor_{int64, double}"


These functions are needed for computing indexing expressions.

We were previously calling into
`aten::native::div_floor_{integer,floating}` when compiling the cpp
wrapper code in non-ABI-compatible mode. However, these functions were
not available when compiling under ABI-compatible mode as we did not
include ATen/native/BinaryOps.h. It is technically safe to call those
functions in ABI-compatible mode as they are header-only, but I didn't
want to include BinaryOps.h into the cpp wrapper file because it exposes
other potentially non-ABI-stable declarations.

As such, I have created a new torch/csrc/inductor/math.h file that
contains versions of div_floor_* that are clearly uncoupled from libc10
and libtorch. We only need to care about int64_t and double types when
computing indexing expressions, so I have only defined
div_floor_{int64_, double}. This makes it easy to decouple from c10
code.

I also noticed that when compiling the cpp kernels, we include
BinaryOps.h directly. I have swapped that out for the new math.h to
better enforce the ABI-compatible interface.

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Enable floor_div indexing to work under ABI-compat mode"


Previously, floor_div operations were defined in
ATen/native/BinaryOps.h. Since this header was not included under
ABI-compat mode, trying to use those indexing operations would result in
compilation errors.

Technically, it is safe to use aten::native::floor_div_* functions in
ABI-compat mode as they are header-only; we could simply include
BinaryOps.h. However, there are other declarations in BinaryOps.h that
are not binary-compatible, so this is not ideal. Thus, I have moved those
functions into a separate file, and put them under c10/util, since they
don't really have tensor-specific logic.

c10 functions are not all header-only, so this still isn't ideal, but
this still seems like an improvement. Moreover, cpp_prefix.h -- used
when compiling cpp kernels -- already includes c10 header files, so
ABI-compatibility already depends on maintaining some c10 functions as
header-only.

[ghstack-poisoned]

int3 commented:
Alternative approach (with rationale in commit mesage): https://github.com/pytorch/pytorch/commit/3d68d92daa9c7a74d845cfbf1c95a22c8e96daaf

I think the current approach is cleaner though.

ezyang commented on torch/csrc/inductor/math.h:
```diff
@@ -0,0 +1,57 @@
+#include <cmath>
+#include "c10/macros/Macros.h"
+
+// Functions in this file should be header-only as it is used in ABI-compatible mode.
```
I don't understand. Is this OK to run from the DSO? Why does it include Macros.h then?

int3 replied:
Macros.h doesn't include anything that links to external C++ code, so it's safe

desertfire replied:
Your statement is true, but it also felt like a slippery slope to me. People who touch those recursively included heads won't necessarily know the ABI constraint of AOTI. Say someone decides to split a header into .h and .cpp, and then it will break our assumption here.

With that being said, I am ok with the current approach if you can also update the linker flags for the abi_compatible mode, e.g.  https://github.com/pytorch/pytorch/blob/cb233dada43f016c8fa58979838939c1adda416b/torch/_inductor/codecache.py#L1301C24-L1301C24. Then if the aforementioned change happens, at least we should be able to catch it with our abi_compatible test cases.

desertfire replied:
Hmm, on a second thought, my suggestion may not be trivial to implement, because the C shim layout does live in libtorch so we need to link against that.

int3 replied:
> Your statement is true, but it also felt like a slippery slope to me.

Yes, I agree that the situation is not ideal. That said I don't think I am actually making the situation that much worse, because cpp_prefix.h already includes a bunch of c10 and aten headers directly (see updated message at the top of the PR.)

> With that being said, I am ok with the current approach if you can also update the linker flags for the abi_compatible mode

It looks like we already avoid linking against c10 when building in fbcode, so I think we're good? I guess it would be better if we checked for `abi_compatible` instead of `is_fbcode`, but it seems they are being used interchangeably throughout the codebase: https://github.com/pytorch/pytorch/blob/cb233dada43f016c8fa58979838939c1adda416b/torch/_inductor/codecache.py#L1398-L1399

(Side note: we are commenting on the "make an alternative implementation of `div_floor_*` functions" approach, which is not the latest version of this PR; the latest version just relocates the existing `div_floor_*` functions from ATen to c10. Either way, the non-linking of c10 makes this safe.)

desertfire replied:
> pp_prefix.h already includes a bunch of c10 and aten headers directly

This should be fixed.

> I guess it would be better if we checked for abi_compatible instead of is_fbcode, but it seems they are being used interchangeably throughout the codebase...

The long-term goal is to only keep the abi-compatible mode for OSS as well. In this case, please change it to check `abi_compatible

int3 committed:
Update on "[inductor] Enable floor_div indexing to work under ABI-compat mode"


Previously, floor_div operations were defined in
ATen/native/BinaryOps.h. Since this header was not included under
ABI-compat mode, trying to use those indexing operations would result in
compilation errors.

Technically, it is safe to use aten::native::floor_div_* functions in
ABI-compat mode as they are header-only; we could simply include
BinaryOps.h. However, there are other declarations in BinaryOps.h that
are not binary-compatible, so this is not ideal. Thus, I have moved those
functions into a separate file, and put them under c10/util, since they
don't really have tensor-specific logic.

c10 functions are not all header-only, so this still isn't ideal, but
this still seems like an improvement. Moreover, cpp_prefix.h -- used
when compiling cpp kernels -- already includes c10 header files, so
ABI-compatibility already depends on maintaining some c10 functions as
header-only.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Enable floor_div indexing to work under ABI-compat mode"


Previously, floor_div operations were defined in
ATen/native/BinaryOps.h. Since this header was not included under
ABI-compat mode, trying to use those indexing operations would result in
compilation errors.

Technically, it is safe to use aten::native::floor_div_* functions in
ABI-compat mode as they are header-only; we could simply include
BinaryOps.h. However, there are other declarations in BinaryOps.h that
are not binary-compatible, so this is not ideal. Thus, I have moved those
functions into a separate file, and put them under c10/util, since they
don't really have tensor-specific logic.

c10 functions are not all header-only, so this still isn't ideal, but
this still seems like an improvement. Moreover, cpp_prefix.h -- used
when compiling cpp kernels -- already includes c10 header files, so
ABI-compatibility already depends on maintaining some c10 functions as
header-only.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Enable floor_div indexing to work under ABI-compat mode"


Previously, floor_div operations were defined in
ATen/native/BinaryOps.h. Since this header was not included under
ABI-compat mode, trying to use those indexing operations would result in
compilation errors.

Technically, it is safe to use aten::native::floor_div_* functions in
ABI-compat mode as they are header-only; we could simply include
BinaryOps.h. However, there are other declarations in BinaryOps.h that
are not binary-compatible, so this is not ideal. Thus, I have moved those
functions into a separate file, and put them under c10/util, since they
don't really have tensor-specific logic.

c10 functions are not all header-only, so this still isn't ideal, but
this still seems like an improvement. Moreover, cpp_prefix.h -- used
when compiling cpp kernels -- already includes c10 header files, so
ABI-compatibility already depends on maintaining some c10 functions as
header-only.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 commented:
@pytorchbot merge

int3 committed:
Update on "[inductor] Enable floor_div indexing to work under ABI-compat mode"


Previously, floor_div operations were defined in
ATen/native/BinaryOps.h. Since this header was not included under
ABI-compat mode, trying to use those indexing operations would result in
compilation errors.

Technically, it is safe to use aten::native::floor_div_* functions in
ABI-compat mode as they are header-only; we could simply include
BinaryOps.h. However, there are other declarations in BinaryOps.h that
are not binary-compatible, so this is not ideal. Thus, I have moved those
functions into a separate file, and put them under c10/util, since they
don't really have tensor-specific logic.

c10 functions are not all header-only, so this still isn't ideal, but
this still seems like an improvement. Moreover, cpp_prefix.h -- used
when compiling cpp kernels -- already includes c10 header files, so
ABI-compatibility already depends on maintaining some c10 functions as
header-only.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 commented:
@pytorchbot merge

vfdev-5 commented:
Hey @int3, I think there could be an error comping from this PR. Here is a repro code and the error message:
```python
import torch


def func(x, a):
    n = (a * 1.0) // 8.0
    y = x + n
    return y

cfunc = torch.compile(func, dynamic=True, fullgraph=True)

device = "cpu"
x = torch.tensor(0, dtype=torch.float32, device=device)
a = 12

out = cfunc(x, a)
expected = func(x, a)
torch.testing.assert_close(out, expected)
```
and
```
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
CppCompileError: C++ compile error

Command:
g++ /tmp/torchinductor_root/cn/ccniuvr7xihkubwhdhaaxiiklktdhdbkijsrbb4jkchgiscpeqxn.cpp -shared -fPIC -Wall -std=c++17 -Wno-unused-variable -Wno-unknown-pragmas -D_GLIBCXX_USE_CXX11_ABI=1 -I/pytorch/torch/include -I/pytorch/torch/include/torch/csrc/api/include -I/pytorch/torch/
include/TH -I/pytorch/torch/include/THC -I/usr/include/python3.8 -L/pytorch/torch/lib -L/usr/lib -L/pytorch/torch/lib -ltorch -ltorch_cpu -lgomp -ltorch_python -lc10 -mavx2 -mfma -DCPU_CAPABILITY_AVX2 -O3 -DNDEBUG -ffast-math -fno-finite-math-only -fno-unsafe-math-optimizations
 -march=native -fopenmp -D C10_USING_CUSTOM_GENERATED_MACROS -o /tmp/torchinductor_root/cn/ccniuvr7xihkubwhdhaaxiiklktdhdbkijsrbb4jkchgiscpeqxn.so

Output:
In file included from /tmp/torchinductor_root/cn/ccniuvr7xihkubwhdhaaxiiklktdhdbkijsrbb4jkchgiscpeqxn.cpp:2:
/tmp/torchinductor_root/26/c26eqbkuxvn72gf7p2xujmqjcwf4bo6lxmp6rwborxnf4gldnimh.h: In function ‘at::vec::CPU_CAPABILITY::Vectorized<float> to_float_mask(int)’:
/tmp/torchinductor_root/26/c26eqbkuxvn72gf7p2xujmqjcwf4bo6lxmp6rwborxnf4gldnimh.h:407:4: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  407 |   *(uint32_t*)&mask = src ? 0xFFFFFFFF : 0;
      |    ^~~~~~~~~~~~~~~~
In file included from /tmp/torchinductor_root/26/c26eqbkuxvn72gf7p2xujmqjcwf4bo6lxmp6rwborxnf4gldnimh.h:16,
                 from /tmp/torchinductor_root/cn/ccniuvr7xihkubwhdhaaxiiklktdhdbkijsrbb4jkchgiscpeqxn.cpp:2:
/pytorch/torch/include/c10/util/generic_math.h: In instantiation of ‘scalar_t c10::div_floor_floating(scalar_t, scalar_t) [with scalar_t = double]’:
/tmp/torchinductor_root/cn/ccniuvr7xihkubwhdhaaxiiklktdhdbkijsrbb4jkchgiscpeqxn.cpp:9:147:   required from here
/pytorch/torch/include/c10/util/generic_math.h:43:31: error: ‘compat_copysign’ was not declared in this scope
   43 |     floordiv = compat_copysign(scalar_t(0), a / b);
      |                ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information

You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True
```

With `cat /tmp/torchinductor_root/cn/ccniuvr7xihkubwhdhaaxiiklktdhdbkijsrbb4jkchgiscpeqxn.cpp`:
```c++
#include "/tmp/torchinductor_root/26/c26eqbkuxvn72gf7p2xujmqjcwf4bo6lxmp6rwborxnf4gldnimh.h"
extern "C" void kernel(const float* in_ptr0,
                       float* out_ptr0,
                       const long ks0)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        auto tmp1 = c10::convert<float>(c10::div_floor_floating(static_cast<double>((1.00000000000000*ks0)), static_cast<double>(8.00000000000000)));
        auto tmp2 = decltype(tmp0)(tmp0 + tmp1);
        out_ptr0[static_cast<long>(0L)] = tmp2;
    }
}
```
